File: Week 6/sentimental-credit/credit.py
from cs50 import get_string, get_int
import re


visa = re.compile(r'4')
amex = re.compile(r'(37|34)')
mastercard = re.compile(r'(51|52|53|54|55)')


def main():
    number = get_int("Number: ")
    if (check_luhn(number)):
        get_credit_card_name(number)
    else:
        print("INVALID")

def check_luhn(n):
    f = 1
    odd_sum = 0
    even_sum = 0
    sum = 0
    while True:
        if f == 1:
            odd_sum += n % 10
            f = 0
        else:
            even_number = n % 10 * 2
            if (even_number > 9):
                even_sum += even_number % 10 + even_number // 10
            else:
                even_sum += even_number
            f = 1
        n //= 10
        if (not n):
            break

    sum = odd_sum + even_sum

    if (sum % 10 == 0):
        return True
    else:
        return False

def get_credit_card_name(n):
    card_number_str = str(n)
    length_of_card = len(card_number_str)
    if (visa.match(card_number_str) != None and (length_of_card == 13 or length_of_card == 16)):
        print("VISA")
    elif (amex.match(card_number_str) != None and length_of_card == 15):
        print("AMEX")
    elif (mastercard.match(card_number_str) != None and length_of_card == 16):
        print("MASTERCARD")
    else:
        print("INVALID")

main()

File: Week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int CheckLuhn(long Number);
void GetCreditCardName(long Number, int f);

int main(void)
{
    long CreditNumber = get_long("Number: \n");
    int f = CheckLuhn(CreditNumber);
    GetCreditCardName(CreditNumber, f);
}

int CheckLuhn(long Number)
{
    int f = 1;
    int odd_sum = 0;
    int even_sum = 0;
    int sum = 0;
    do
    {
        if (f == 1)
        {
            odd_sum += Number % 10;
            f = 0;
        }
        else
        {
            int even_number = Number % 10 * 2;
            if (even_number > 9)
            {
                even_sum += even_number % 10 + even_number / 10;
            }
            else
            {
                even_sum += even_number;
            }
            f = 1;
        }
        Number /= 10;
    }
    while (Number > 0);
    sum = odd_sum + even_sum;
    if (sum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

void GetCreditCardName(long Number, int f)
{
    if (f == 0)
    {
        printf("INVALID\n");
    }
    else
    {
        int first_number;
        if (Number > 999999999999999)
        {
            first_number = Number / 1000000000000000;
            if (first_number == 4)
            {
                printf("VISA\n");
            }
            else
            {
                first_number = Number / 100000000000000;
                if (first_number == 51 || first_number == 52 || first_number == 53 ||
                    first_number == 54 || first_number == 55)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
        }
        else if (Number > 99999999999999 && Number < 1000000000000000)
        {
            first_number = Number / 10000000000000;
            if (first_number == 34 || first_number == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (Number > 999999999999 && Number < 10000000000000)
        {
            first_number = Number / 1000000000000;
            if (first_number == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}
